src/scenarios/shared/runner.py

```diff
@@ -89,114 +66,100 @@ def parseargs(self):
+                             )
+
         elif self.testtype == const.SDK:
             envlistbuild = 'DOTNET_MULTILEVEL_LOOKUP=0'
-            envlistcleanbuild= ';'.join(['MSBUILDDISABLENODEREUSE=1', envlistbuild])
+            envlistcleanbuild = ';'.join(['MSBUILDDISABLENODEREUSE=1', envlistbuild])
             # clean build
             if self.sdktype == const.CLEAN_BUILD:
-                startup.runtests(scenarioname=self.scenarioname,
-                                exename=self.traits.exename,
-                                guiapp=self.traits.guiapp,
-                                startupmetric=const.STARTUP_PROCESSTIME,
-                                appargs='build',
-                                timeout=self.traits.timeout,
-                                warmup='true',
-                                iterations=self.traits.iterations,
-                                scenariotypename='%s_%s' % (const.SCENARIO_NAMES[const.SDK], const.CLEAN_BUILD),
-                                apptorun=const.DOTNET,
-                                iterationsetup='py' if sys.platform == 'win32' else 'py3',
-                                setupargs='-3 %s setup_build' % const.ITERATION_SETUP_FILE if sys.platform == 'win32' else const.ITERATION_SETUP_FILE,
-                                iterationcleanup='py' if sys.platform == 'win32' else 'py3',
-                                cleanupargs='-3 %s cleanup' % const.ITERATION_SETUP_FILE if sys.platform == 'win32' else const.ITERATION_SETUP_FILE,
-                                workingdir= const.APPDIR if not self.traits.workingdir else os.path.join(const.APPDIR, self.traits.workingdir),
-                                environmentvariables=envlistcleanbuild,
-                                processwillexit=self.traits.processwillexit,
-                                measurementdelay=self.traits.measurementdelay
-                             )
+                self.traits.add_traits(
+                    overwrite=False,
+                    appargs='build',
+                    iterationsetup=pythoncommand(),
+                    setupargs='%s %s setup_build' % ('-3' if iswin() else '', const.ITERATION_SETUP_FILE),
+                    iterationcleanup=pythoncommand(),
+                    cleanupargs='%s %s cleanup' % ('-3' if iswin() else '', const.ITERATION_SETUP_FILE),
+                    workingdir=const.APPDIR,
+                    environmentvariables=envlistcleanbuild,
+                )
+                self.traits.add_traits(overwrite=True, startupmetric=const.STARTUP_PROCESSTIME)
+                startup.runtests(**self.traits.get_all_traits(),
+                                 scenarioname=self.scenarioname,
+                                 scenariotypename='%s_%s' % (const.SCENARIO_NAMES[const.SDK], const.CLEAN_BUILD),
+                                 apptorun=const.DOTNET
+                                 )
```

Comment:
Yes this is the same question that constantly bothers me. My understanding is ```TestTrait``` should be attributes that are asset-specific, and anything scenario-specific should go to ```runtests()```. The reason for this classification is based on our design --- we create ```TestTraits``` in ```test.py```, which exists individually in each asset, and we override them in ```runner.py``` for each scenario . Thus, we want to put attributes that not only belong to this asset but also are MOST SHARED across different scenario in ```TestTraits```, so we don't have to override them in different scenarios, otherwise they are just a group of kwargs. For example, I think ```appargs``` is a bad test trait, because we very possibly have different ```appargs``` across different scenarios, so is ```workingdir```, ```iterationsetup```, etc...

Reply:
Ah. Got it. Now I understand the confusion.

My intent with the traits was to have it be the holding area for everything required to run a test. That includes app args, working directories, everything. If you go back to the first [checkin](https://github.com/dotnet/performance/commit/b8483afd517af2eeb0e0c45018162ecdc6695f88#diff-c701c2cae0458d92a0b5cdab39708506R60-R64) you can kind of see that.

If we want to establish a more precise set of principles that's fine. As we evolved things and the complexity grew perhaps the original concept didn't make sense any longer. Maybe the whole idea of the traits type is bad?

Reply:
Got it. If this is true, kwargs like ```scenariotypename```, ```scenarioname``` and ```apptorun``` seem no less like a trait than the existing traits. Maybe we can just put everything in test traits, let ```startup.runtests()``` choose what it wants, and make ```test.py``` be an access point that's asset-specific and can override the default scenario-specific values?  Also maybe it's good to differentiate which is accessible by ```test.py``` and which is not.